### DISCUSSION TITLE
[macOS] Multi-window app does not terminate

### DISCUSSION THREAD

**Discussion by MartyIX** [dotnet/maui#13574] [General] [answered]
### Description

When I enable multi-window support for MAUI on MacCatalyst according to [official instructions](https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/windows?view=net-maui-7.0#multi-window-support), the application does not terminate when the app's "X" button is clicked.

### Steps to Reproduce

Check out the commit https://github.com/MartyIX/macOS-MAUI-multiwindow/commit/7b4861c80153b3cc7ebd86ec669a9c7104608d34 (vanilla MAUI application):

1. Start the application
2. Click the "X" (red) button to terminate the application
3. Application **does** terminate.

Check out the latest commit of https://github.com/MartyIX/macOS-MAUI-multiwindow (https://github.com/MartyIX/macOS-MAUI-multiwindow/commit/d2ed4aca7ffb091da15914e8d054c7565d5bf1dc) that adds the multi-window support:

1. Start the application
2. Click the "X" (red) button to terminate the application
3. Application **does not** terminate.

### Link to public reproduction project repository

https://github.com/MartyIX/macOS-MAUI-multiwindow

### Version with bug

7.0 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

macOS

### Affected platform versions

macOS Ventura 13.2.1

### Did you find any workaround?

No.

**Comment by davidbuckleyni** [comment]
Usually in .net apps u have to create the close code the apps dont no how to other wise something like Application.Terminate?

**Reply by MartyIX** [reply]
The sentence is hard to parse. Anyway, if some special handling is required for multi-window apps on macOS then it should be documented imho.

**Comment by MartyIX** [comment]
Answer from @drasticactions:

> That's not a MAUI bug. That's typically how macOS Catalyst apps work.
When you close all the open windows, the app will continue to run. Ex. you can click on the dock icon and it will spawn a new window through the Window Scene.
> So you would need to handle the window state and terminate the application if that's what you want.

So to put it simply: You close the last window but that does not automatically means that the application is stopped (as is customary on Windows and Linux).

Anyway, I tried to modify my `Platforms/MacCatalyst/AppDelegate.cs` file as follows:

```diff
using System.Diagnostics.CodeAnalysis;
using Foundation;
using Microsoft.Maui;
using Microsoft.Maui.Hosting;
using UIKit;

namespace macOS_maui_multiwindow;

[Register("AppDelegate")]
public class AppDelegate : MauiUIApplicationDelegate
{
+    /// <summary>
+    /// Returns a boolean value that indicates if the app terminates once the last window closes.
+    /// </summary>
+    /// <param name="application">Application object whose last window was closed.</param>
+    /// <returns>Always <c>true</c>.</returns>
+    /// <seealso href="https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428381-applicationshouldterminateafterl"/>
+    [Export("applicationShouldTerminateAfterLastWindowClosed:")]
+    public virtual bool ApplicationShouldTerminateAfterLastWindowClosed(UIApplication application)
+    {
+       return true;
+    }

    /// <inheritdoc/>
    protected override MauiApp CreateMauiApp() => MauiProgram.CreateMauiApp();
}
```

but unfortunately, it's *not* called because this API works for AppKit and not for UIKit. So this won't work.

Next action is to try: https://github.com/drasticactions/Drastic.Tools/blob/e141a5785537021890a26390c970c40239ae703d/src/Drastic.TrayWindow/NSApplication.Catalyst.cs#L39-L47

Resources
* https://stackoverflow.com/questions/61206065/how-to-terminate-programmatically-a-macos-app-when-clicking-on-the-top-left-red/61206288#61206288
* https://stackoverflow.com/questions/8233141/completely-close-an-os-x-application-with-window-close-button
* https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428381-applicationshouldterminateafterl
* https://github.com/dotnet/maui/blob/04bfb275d9401a631d3e9a1e20665e68e99b79c4/src/Core/src/Platform/iOS/MauiUIApplicationDelegate.cs

**Comment by drasticactions** [comment, marked as answer]
This issue is not MAUI Specific; this is how Mac Catalyst generally works. With a single instance application, closing the window closes the application. When you enable multiple windows, the opposite happens; Closing all of the open windows does not close the application but keeps it running so you can open windows from the dock.

As I pointed out, you can change that behavior by accessing the underlying AppKit Delegate and forcing it to change. Still, these are private APIs and are discouraged from use. 

I am moving this to a discussion as this is not a MAUI bug.

**Comment by ilmian** [comment]
ApplicationShouldTerminateAfterLastWindowClosed should be exposed by MauiUIApplicationDelegate to be overridden in AppDelegate, so we can use its o handle the closing the app via red button. We were able to override it in Xamarin.MacOS's AppDelegate.

**Reply by drasticactions** [reply]
These are AppKit APIs for macOS apps, and are not exposed for Mac Catalyst apps, which use UIKit (Which is what MAUI apps for macOS are based on). You can't do this without hacking the underlying application layer. As I said above, this is not a MAUI issue, it's a limitation of Mac Catalyst apps and you would have the same issue if you built the app straight using Mac Catalyst (in C#, Swift, Obj-C, etc) without MAUI.

So MAUI should never support it at a platform level, as until Apple allows an official way to do it, every other method would be a hack on the underlying platform.

**Comment by Bodkens** [comment]
you can handle window destroying for each window you create, but it is not best solution, although only possible right now

Application window
```

public partial class App : Application
{
 
    protected override Window
    CreateWindow(IActivationState activationState)
    {
        Window window = base.CreateWindow(activationState);
        window.Destroying += (object sender, EventArgs args) =>
        {
            if (Current.Windows.Count - 1 == 0) Application.Current.Quit(); //Or Environment.Exit(0)
          
        };
        return window;
    }
```
    
   
    
 ```
//Another windows you will create
   Window window = new Window(...)
    window.Destroying += (object sender, EventArgs args) =>
        {
            if (Current.Windows.Count - 1 == 0) Application.Current.Quit();
        };
```
